File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
# 프로그래머스 49189 가장 먼 노드 LV.3
from collections import deque


def solution(n, vertex):
    graph = [[] for _ in range(n + 1)]
    for a, b in vertex:
        graph[a].append(b)
        graph[b].append(a)

    visited = [0] * (n + 1)
    q = deque()
    q.append(1)
    while q:
        start = q.popleft()
        for num in graph[start]:
            if visited[num] == 0 and num != 1:
                visited[num] = visited[start] + 1
                q.append(num)
                
    return visited.count(max(visited))

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Node {
    int id, cnt;
    
    Node(int id, int cnt) {
        this.id = id;
        this.cnt = cnt;
    }
}

class Solution {
    int N;
    List<Integer>[] graph;
    
    public int solution(int n, int[][] edge) {
        init(n, edge);
        return bfs(1);
    }
    
    private int bfs(int start) {
        Queue<Node> q = new LinkedList<>();
        boolean[] visited = new boolean[N + 1];
        q.add(new Node(start, 0));
        visited[start] = true;
        
        int maxCnt = Integer.MIN_VALUE, answer = 0;
        while (!q.isEmpty()) {
            Node now = q.poll();
            boolean isLeaf = true;
            
            for (int nxt : graph[now.id]) {
                if (visited[nxt]) continue;
                
                q.add(new Node(nxt, now.cnt + 1));
                visited[nxt] = true;
                isLeaf = false;
            }
            
            if (isLeaf) {
                if (now.cnt > maxCnt) {
                    maxCnt = now.cnt;
                    answer = 1;
                } else if (now.cnt == maxCnt) answer++;
            }
        }
        
        return answer;
    }
    
    private void init(int n, int[][] edge) {
        N = n;
        graph = new List[n + 1];
        for (int i = 1; i <= n; i++) graph[i] = new ArrayList<>();
        for (int[] e : edge) {
            graph[e[0]].add(e[1]);
            graph[e[1]].add(e[0]);
        }
    }
}